src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,389 @@
+// limitations under the License.
+
+use crate::token::{Token, TokenProvider};
+use crate::Result;
+use std::sync::{Arc, RwLock};
+use tokio::sync::{Mutex, Notify};
+// Using tokio's wrapper makes the cache testable without relying on clock times.
+use tokio::time::Instant;
+
+// TODO(#1210) - implementation in progress
+#[allow(dead_code)]
+#[derive(Debug)]
+pub(crate) struct TokenCache<T>
+where
+    T: TokenProvider,
+{
+    // The cached token, or the last seen error.
+    token: Arc<RwLock<Result<Token>>>,
+
+    // Tracks if a refresh is ongoing. If the lock is held, there is a refresh.
+    refresh_in_progress: Arc<Mutex<()>>,
+    // Allows us to await the result of a refresh in multiple threads.
+    refresh_notify: Arc<Notify>,
+
+    // The token provider. This thing does the refreshing.
+    inner: Arc<T>,
+}
+
+// Returns true if we are holding an error, or a token that has expired.
+fn is_invalid(token: &Result<Token>) -> bool {
+    match token {
+        Ok(t) => t.expires_at.is_some_and(|e| e <= Instant::now().into_std()),
+        Err(_) => true,
+    }
+}
+
+// We manually implement the `Clone` trait because the Rust compiler will
+// squawk if `T` is not `Clone`, even though we only hold an `Arc<T>`. :shrug:
+impl<T: TokenProvider> Clone for TokenCache<T> {
+    fn clone(&self) -> TokenCache<T> {
+        TokenCache {
+            token: self.token.clone(),
+            refresh_in_progress: self.refresh_in_progress.clone(),
+            refresh_notify: self.refresh_notify.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<T: TokenProvider> TokenCache<T> {
+    // TODO(#1210) - implementation in progress
+    #[allow(dead_code)]
+    pub fn new(inner: T) -> TokenCache<T> {
+        TokenCache {
+            token: Arc::new(RwLock::new(Err(crate::errors::CredentialError::retryable_from_str("No token in the cache. This should never happen. Something has gone wrong. Open an issue at https://github.com/googleapis/google-cloud-rust/issues/new?template=bug_report.md.")))),
+            refresh_in_progress: Arc::new(Mutex::new(())),
+            refresh_notify: Arc::new(Notify::new()),
+            inner: Arc::new(inner),
+        }
+    }
+
+    // Clones the current token, in a thread-safe manner. Releases the lock on return.
+    async fn current_token(&self) -> Result<Token> {
+        self.token.read().unwrap().clone()
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: TokenProvider + 'static> TokenProvider for TokenCache<T> {
+    async fn get_token(&self) -> Result<Token> {
+        let token = self.current_token().await;
+
+        if !is_invalid(&token) {
+            return token;
+        }
+
+        match self.refresh_in_progress.try_lock() {
+            // Check if there are any outstanding refreshes...
+            Ok(guard) => {
+                // No refreshes. We should start one.
+                let token = self.inner.get_token().await;
```

Comment:
You seem to have a race condition here.  You grab and release the token, then check if there is a refresh.  Some other task / thread could have updated the token after like 83 and before like 89. That is extra work that you could eliminate.

Reply:
I agree.

Reply:
I think my problem is that the state of the token cache is split across `refresh_in_progress` and `token`. And I try to update them one at a time (transitioning through an unwanted state).

I need something new to hold state and manage the transitions between states.

Reply:
Ok, I think there is a simpler fix. I just need to hold the token lock when I try the refresh lock.

That will prevent a refresh from writing the token behind my back.

Thanks.

Reply:
Fixed this problem.

I think the one other thing that needs scrutiny is the `notify` business.

Maybe I fail to acquire the lock guard. Then we could release the lock and notify all waiters:

https://github.com/googleapis/google-cloud-rust/blob/d044da84ccb96aff14eedc837b0a4f85eda1c90b/src/auth/src/token_cache.rs#L121-L125

before we await the notify:

https://github.com/googleapis/google-cloud-rust/blob/d044da84ccb96aff14eedc837b0a4f85eda1c90b/src/auth/src/token_cache.rs#L132

Hm. I think this is a problem.

Reply:
Thinking out loud...

If I fail to acquire the refresh lock, instead of waiting for some notify business, I could just demand the refresh lock.

When I am given the lock, it means the refresh operation has completed. Then I just return whatever is in my cache.

Reply:
I fixed this.